Rootless podman containers cause glances to fail with KeyError

**Check the bug**
Before filling this bug report, please search if a similar issue already exists.
In this case, just add a comment on this existing issue.

**Describe the bug**
A clear and concise description of what the bug is.

Displaying container status for rootless podman is not working, instead i get following error, when exiting:
```
Exception in thread Thread-7 (__update_plugin):
Traceback (most recent call last):
  File "/usr/lib64/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.12/threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/stats.py", line 258, in __update_plugin
    self._plugins[p].update()
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/plugin/model.py", line 1151, in wrapper
    ret = fct(self, *args, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/plugin/model.py", line 1168, in wrapper
    ret = fct(*args, **kw)
          ^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/__init__.py", line 223, in update
    stats_podman = self.update_podman() if self.podman_extension else {}
                   ^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/__init__.py", line 245, in update_podman
    version, containers = self.podman_extension.update(all_tag=self._all_tag())
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/engines/podman.py", line 290, in update
    container_stats = [self.generate_stats(container) for container in containers]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/share/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/engines/podman.py", line 328, in generate_stats
    stats['cpu_percent'] = stats["cpu"]['total']
                           ~~~~~~~~~~~~^^^^^^^^^
KeyError: 'total'
```


**To Reproduce**
1. create this simple config file:

```
[containers]
disable=False
all=True
podman_sock=unix:///run/user/1000/podman/podman.sock
```

2. Start Glances with the following options `glances -C glances.ini`
3. See error

**Expected behavior**
The glances will display running podman containers, and not errors

**Environement (please complete the following information)**
 - Operating System (lsb_release -a or OS name/version): `To be completed with result of:
```
LSB Version:	n/a
Distributor ID:	Fedora
Description:	Fedora Linux 40 (Forty)
Release:	40
Codename:	n/a
```
 - Glances & psutil versions: `To be completed with result of:
 - 
```
Glances version:	4.0.8
Glances API version:	4
PsUtil version:		5.9.8
Log file:		/home/tokariew/.cache/glances/glances.log
```
 - How do you install Glances (Pypi package, script, package manager, source): pipx
 - Glances test: ` To be completed with result of: glances --issue`

```
➜ glances --issue -C ~/glances.ini
Exception in thread Thread-7 (__update_plugin):
Traceback (most recent call last):
  File "/usr/lib64/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.12/threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/stats.py", line 258, in __update_plugin
    self._plugins[p].update()
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/plugin/model.py", line 1151, in wrapper
    ret = fct(self, *args, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/plugin/model.py", line 1168, in wrapper
    ret = fct(*args, **kw)
          ^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/__init__.py", line 223, in update
    stats_podman = self.update_podman() if self.podman_extension else {}
                   ^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/__init__.py", line 245, in update_podman
    version, containers = self.podman_extension.update(all_tag=self._all_tag())
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/engines/podman.py", line 290, in update
    container_stats = [self.generate_stats(container) for container in containers]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tokariew/.local/pipx/venvs/glances/lib64/python3.12/site-packages/glances/plugins/containers/engines/podman.py", line 328, in generate_stats
    stats['cpu_percent'] = stats["cpu"]['total']
                           ~~~~~~~~~~~~^^^^^^^^^
KeyError: 'total'
===============================================================================
Glances 4.0.8 (/home/tokariew/.local/pipx/venvs/glances/lib/python3.12/site-packages/glances/__init__.py)
Python 3.12.3 (/home/tokariew/.local/pipx/venvs/glances/bin/python)
PsUtil 5.9.8 (/home/tokariew/.local/pipx/venvs/glances/lib/python3.12/site-packages/psutil/__init__.py)
===============================================================================
alert         [OK]    0.00001s 
[]
amps          [OK]    0.00006s 
[]
cloud         [OK]    0.00003s 
{}
connections   [OK]    0.05131s 
{'ESTABLISHED': 24,
 'LISTEN': 17,
 'SYN_RECV': 0,
 'SYN_SENT': 2,
 'initiated': 2,
 'net_connections_enabled': True,
 'nf_conntrack_count': 53.0,
 'nf_conntrack_enabled': True,
 'nf_conntrack_max': 262144.0,
 'nf_conntrack_percent': 0.0202178955078125,
 'terminated': 2}
containers    [ERROR] 0.03413s 'total'
core          [OK]    0.00121s 
{'log': 12, 'phys': 6}
cpu           [OK]    0.00058s 
{'cpucore': 12,
 'ctx_switches': 0,
 'ctx_switches_gauge': 4375679908,
 'guest': 0.0,
 'idle': 69.9,
 'interrupts': 0,
 'interrupts_gauge': 1711358773,
 'iowait': 1.0,
 'irq': 0.7,
 'nice': 0.0,
 'soft_interrupts': 0,
 'soft_interrupts_gauge': 504551810,
 'steal': 0.0,
 'syscalls': 0,
 'syscalls_gauge': 0,
 'system': 4.9,
 'time_since_update': 2.281040668487549,
 'total': 50.0,
 'user': 23.3}
diskio        [OK]    0.00094s key=disk_name 
[{'disk_name': 'nvme0n1',
  'key': 'disk_name',
  'read_bytes': 0,
  'read_bytes_gauge': 46388284416,
  'read_count': 0,
  'read_count_gauge': 1543583,
  'time_since_update': 2.283613681793213,
  'write_bytes': 0,
  'write_bytes_gauge': 72489518080,
  'write_count': 0,
  'write_count_gauge': 3493038}, ...]
folders       [OK]    0.00004s 
[]
fs            [OK]    0.00100s key=mnt_point 
[{'device_name': '/dev/mapper/luks-874950f7-5c87-4d70-9abb-1e08ae3aca76',
  'free': 627858595840,
  'fs_type': 'btrfs',
  'key': 'mnt_point',
  'mnt_point': '/',
  'percent': 36.8,
  'size': 998483427328,
  'used': 365083500544}, ...]
gpu           [OK]    0.01921s key=gpu_id 
[{'fan_speed': None, 'gpu_id': 'amd0', 'key': 'gpu_id', 'mem': 19, 'name': 'AMD GPU', 'proc': 89, 'temperature': 77}, ...]
help          [OK]    0.00001s 
None
ip            [OK]    0.00004s 
{}
irq           [OK]    0.00148s key=irq_line 
[{'irq_line': 'CAL', 'irq_rate': 60252, 'key': 'irq_line', 'time_since_update': 2.3028995990753174}, ...]
load          [OK]    0.00005s 
{'cpucore': 12, 'min1': 1.21240234375, 'min15': 1.25048828125, 'min5': 1.2548828125}
mem           [OK]    0.00025s 
{'active': 4660936704,
 'available': 27342598144,
 'buffers': 3592192,
 'cached': 18195009536,
 'free': 27342598144,
 'inactive': 15017549824,
 'percent': 18.5,
 'shared': 285134848,
 'total': 33557012480,
 'used': 6214414336}
memswap       [OK]    0.00040s 
{'free': 15944380416,
 'percent': 7.2,
 'sin': 790790144,
 'sout': 2494521344,
 'time_since_update': 2.3022847175598145,
 'total': 17179865088,
 'used': 1235484672}
network       [OK]    0.00207s key=interface_name 
[{'alias': None,
  'bytes_all': 6972,
  'bytes_all_gauge': 112191242700,
  'bytes_all_rate_per_sec': 3295.0,
  'bytes_recv': 3486,
  'bytes_recv_gauge': 56095621350,
  'bytes_recv_rate_per_sec': 1647.0,
  'bytes_sent': 3486,
  'bytes_sent_gauge': 56095621350,
  'bytes_sent_rate_per_sec': 1647.0,
  'interface_name': 'lo',
  'key': 'interface_name',
  'speed': 0,
  'time_since_update': 2.115530252456665}, ...]
now           [OK]    0.00005s 
{'custom': '2024-06-10 13:11:38 CEST', 'iso': '2024-06-10T13:11:38+02:00'}
percpu        [OK]    0.00017s key=cpu_number 
[{'cpu_number': 0,
  'guest': 0.0,
  'guest_nice': 0.0,
  'idle': 0.0,
  'iowait': 0.0,
  'irq': 0.0,
  'key': 'cpu_number',
  'nice': 0.0,
  'softirq': 0.0,
  'steal': 0.0,
  'system': 0.0,
  'total': 100.0,
  'user': 0.0}, ...]
ports         [OK]    0.00027s 
[]
processcount  [OK]    0.10009s 
{'pid_max': 0, 'running': 2, 'sleeping': 229, 'thread': 1028, 'total': 422}
processlist   [OK]    0.00038s 
[]
psutilversion [OK]    0.00002s 
'5.9.8'
quicklook     [OK]    0.00044s 
{'cpu': 50.0,
 'cpu_hz': 3600000000.0,
 'cpu_hz_current': 3859061583.3333325,
 'cpu_log_core': 12,
 'cpu_name': 'AMD Ryzen 5 3600 6-Core Processor',
 'cpu_phys_core': 6,
 'load': 10.4,
 'mem': 18.5,
 'percpu': [{...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}],
 'swap': 7.2}
raid          [OK]    0.00002s 
{}
sensors       [OK]    0.00001s key=label 
[{'critical': 115,
  'key': 'label',
  'label': 'AUXTIN0',
  'type': <SensorType.CPU_TEMP: 'temperature_core'>,
  'unit': 'C',
  'value': 61,
  'warning': 115}, ...]
smart         [NA]
system        [OK]    0.00001s 
{'hostname': 'ryzen.tokariew.xyz',
 'hr_name': 'Fedora Linux 40 64bit / Linux 6.8.11-300.fc40.x86_64',
 'linux_distro': 'Fedora Linux 40',
 'os_name': 'Linux',
 'os_version': '6.8.11-300.fc40.x86_64',
 'platform': '64bit'}
uptime        [OK]    0.00013s 
{'seconds': 98459}
version       [OK]    0.00002s 
'4.0.8'
wifi          [OK]    0.00010s 
[]
===============================================================================
Total time to update all stats: 0.21884s
===============================================================================
```

**Additional context**
version reported by pipx:
```
   package glances 4.0.8, installed using Python 3.12.3
    - glances
    - man1/glances.1
    Injected Packages:
      - fastapi 0.111.0
      - podman 5.0.0
```

You can also [pastebin](https://pastebin.com/):

* the Glances configuration file (https://glances.readthedocs.io/en/latest/config.html#location)
* the Glances log file (https://glances.readthedocs.io/en/latest/config.html#logging)

--- a/glances/plugins/containers/engines/docker.py
+++ b/glances/plugins/containers/engines/docker.py
@@ -9,10 +9,11 @@
 """Docker Extension unit for Glances' Containers plugin."""
 
 import time
+from typing import Any, Dict, List, Optional, Tuple
 
 from glances.globals import iterkeys, itervalues, nativestr, pretty_date, replace_special_chars
 from glances.logger import logger
-from glances.plugins.containers.stats_streamer import StatsStreamer
+from glances.plugins.containers.stats_streamer import ThreadedIterableStreamer
 
 # Docker-py library (optional and Linux-only)
 # https://github.com/docker/docker-py
@@ -43,7 +44,7 @@ class DockerStatsFetcher:
 
         # Threaded Streamer
         stats_iterable = container.stats(decode=True)
-        self._streamer = StatsStreamer(stats_iterable, initial_stream_value={})
+        self._streamer = ThreadedIterableStreamer(stats_iterable, initial_stream_value={})
 
     def _log_debug(self, msg, exception=None):
         logger.debug(f"containers (Docker) ID: {self._container.id} - {msg} ({exception}) ")
@@ -53,7 +54,7 @@ class DockerStatsFetcher:
         self._streamer.stop()
 
     @property
-    def activity_stats(self):
+    def activity_stats(self) -> Dict[str, Dict[str, Any]]:
         """Activity Stats
 
         Each successive access of activity_stats will cause computation of activity_stats
@@ -63,7 +64,7 @@ class DockerStatsFetcher:
         self._last_stats_computed_time = time.time()
         return computed_activity_stats
 
-    def _compute_activity_stats(self):
+    def _compute_activity_stats(self) -> Dict[str, Dict[str, Any]]:
         with self._streamer.result_lock:
             io_stats = self._get_io_stats()
             cpu_stats = self._get_cpu_stats()
@@ -78,11 +79,11 @@ class DockerStatsFetcher:
         }
 
     @property
-    def time_since_update(self):
+    def time_since_update(self) -> float:
         # In case no update, default to 1
         return max(1, self._streamer.last_update_time - self._last_stats_computed_time)
 
-    def _get_cpu_stats(self):
+    def _get_cpu_stats(self) -> Optional[Dict[str, float]]:
         """Return the container CPU usage.
 
         Output: a dict {'total': 1.49}
@@ -116,7 +117,7 @@ class DockerStatsFetcher:
         # Return the stats
         return stats
 
-    def _get_memory_stats(self):
+    def _get_memory_stats(self) -> Optional[Dict[str, float]]:
         """Return the container MEMORY.
 
         Output: a dict {'usage': ..., 'limit': ..., 'inactive_file': ...}
@@ -139,7 +140,7 @@ class DockerStatsFetcher:
         # Return the stats
         return stats
 
-    def _get_network_stats(self):
+    def _get_network_stats(self) -> Optional[Dict[str, float]]:
         """Return the container network usage using the Docker API (v1.0 or higher).
 
         Output: a dict {'time_since_update': 3000, 'rx': 10, 'tx': 65}.
@@ -168,7 +169,7 @@ class DockerStatsFetcher:
         # Return the stats
         return stats
 
-    def _get_io_stats(self):
+    def _get_io_stats(self) -> Optional[Dict[str, float]]:
         """Return the container IO usage using the Docker API (v1.0 or higher).
 
         Output: a dict {'time_since_update': 3000, 'ior': 10, 'iow': 65}.
@@ -221,7 +222,7 @@ class DockerContainersExtension:
 
         self.connect()
 
-    def connect(self):
+    def connect(self) -> None:
         """Connect to the Docker server."""
         # Init the Docker API Client
         try:
@@ -236,12 +237,12 @@ class DockerContainersExtension:
         # return self.client.version()
         return {}
 
-    def stop(self):
+    def stop(self) -> None:
         # Stop all streaming threads
         for t in itervalues(self.stats_fetchers):
             t.stop()
 
-    def update(self, all_tag):
+    def update(self, all_tag) -> Tuple[Dict, List[Dict]]:
         """Update Docker stats using the input method."""
 
         if not self.client:
@@ -280,22 +281,30 @@ class DockerContainersExtension:
         return version_stats, container_stats
 
     @property
-    def key(self):
+    def key(self) -> str:
         """Return the key of the list."""
         return 'name'
 
-    def generate_stats(self, container):
+    def generate_stats(self, container) -> Dict[str, Any]:
         # Init the stats for the current container
         stats = {
             'key': self.key,
-            # Export name
             'name': nativestr(container.name),
-            # Container Id
             'id': container.id,
-            # Container Status (from attrs)
             'status': container.attrs['State']['Status'],
             'created': container.attrs['Created'],
             'command': [],
+            'io': {},
+            'cpu': {},
+            'memory': {},
+            'network': {},
+            'io_rx': None,
+            'io_wx': None,
+            'cpu_percent': None,
+            'memory_percent': None,
+            'network_rx': None,
+            'network_tx': None,
+            'uptime': None,
         }
 
         # Container Image
@@ -312,37 +321,31 @@ class DockerContainersExtension:
         if not stats['command']:
             stats['command'] = None
 
-        if stats['status'] in self.CONTAINER_ACTIVE_STATUS:
-            started_at = container.attrs['State']['StartedAt']
-            stats_fetcher = self.stats_fetchers[container.id]
-            activity_stats = stats_fetcher.activity_stats
-            stats.update(activity_stats)
+        if stats['status'] not in self.CONTAINER_ACTIVE_STATUS:
+            return stats
 
-            # Additional fields
-            stats['cpu_percent'] = stats["cpu"]['total']
-            stats['memory_usage'] = stats["memory"].get('usage')
-            if stats['memory'].get('cache') is not None:
-                stats['memory_usage'] -= stats['memory']['cache']
-            if 'time_since_update' in stats['io']:
-                stats['io_rx'] = stats['io'].get('ior') // stats['io'].get('time_since_update')
-                stats['io_wx'] = stats['io'].get('iow') // stats['io'].get('time_since_update')
-            if 'time_since_update' in stats['network']:
-                stats['network_rx'] = stats['network'].get('rx') // stats['network'].get('time_since_update')
-                stats['network_tx'] = stats['network'].get('tx') // stats['network'].get('time_since_update')
-            stats['uptime'] = pretty_date(parser.parse(started_at).astimezone(tz.tzlocal()).replace(tzinfo=None))
-            # Manage special chars in command (see isse#2733)
-            stats['command'] = replace_special_chars(' '.join(stats['command']))
-        else:
-            stats['io'] = {}
-            stats['cpu'] = {}
-            stats['memory'] = {}
-            stats['network'] = {}
-            stats['io_rx'] = None
-            stats['io_wx'] = None
-            stats['cpu_percent'] = None
-            stats['memory_percent'] = None
-            stats['network_rx'] = None
-            stats['network_tx'] = None
-            stats['uptime'] = None
+        stats_fetcher = self.stats_fetchers[container.id]
+        activity_stats = stats_fetcher.activity_stats
+        stats.update(activity_stats)
+
+        # Additional fields
+        stats['cpu_percent'] = stats['cpu']['total']
+        stats['memory_usage'] = stats['memory'].get('usage')
+        if stats['memory'].get('cache') is not None:
+            stats['memory_usage'] -= stats['memory']['cache']
+
+        if all(k in stats['io'] for k in ('ior', 'iow', 'time_since_update')):
+            stats['io_rx'] = stats['io']['ior'] // stats['io']['time_since_update']
+            stats['io_wx'] = stats['io']['iow'] // stats['io']['time_since_update']
+
+        if all(k in stats['network'] for k in ('rx', 'tx', 'time_since_update')):
+            stats['network_rx'] = stats['network']['rx'] // stats['network']['time_since_update']
+            stats['network_tx'] = stats['network']['tx'] // stats['network']['time_since_update']
+
+        started_at = container.attrs['State']['StartedAt']
+        stats['uptime'] = pretty_date(parser.parse(started_at).astimezone(tz.tzlocal()).replace(tzinfo=None))
+
+        # Manage special chars in command (see issue#2733)
+        stats['command'] = replace_special_chars(' '.join(stats['command']))
 
         return stats

--- a/glances/plugins/containers/engines/podman.py
+++ b/glances/plugins/containers/engines/podman.py
@@ -7,11 +7,13 @@
 
 """Podman Extension unit for Glances' Containers plugin."""
 
+import time
 from datetime import datetime
+from typing import Any, Dict, Optional, Tuple
 
 from glances.globals import iterkeys, itervalues, nativestr, pretty_date, replace_special_chars, string_value_to_float
 from glances.logger import logger
-from glances.plugins.containers.stats_streamer import StatsStreamer
+from glances.plugins.containers.stats_streamer import ThreadedIterableStreamer
 
 # Podman library (optional and Linux-only)
 # https://pypi.org/project/podman/
@@ -26,63 +28,94 @@ else:
 
 
 class PodmanContainerStatsFetcher:
-    MANDATORY_FIELDS = ["CPU", "MemUsage", "MemLimit", "NetInput", "NetOutput", "BlockInput", "BlockOutput"]
+    MANDATORY_FIELDS = ["CPU", "MemUsage", "MemLimit", "BlockInput", "BlockOutput"]
 
     def __init__(self, container):
         self._container = container
 
+        # Previous stats are stored in the self._old_computed_stats variable
+        # We store time data to enable rate calculations to avoid complexity for consumers of the APIs exposed.
+        self._old_computed_stats = {}
+
+        # Last time when output stats (results) were computed
+        self._last_stats_computed_time = 0
+
         # Threaded Streamer
         stats_iterable = container.stats(decode=True)
-        self._streamer = StatsStreamer(stats_iterable, initial_stream_value={})
-
-    def _log_debug(self, msg, exception=None):
-        logger.debug(f"containers (Podman) ID: {self._container.id} - {msg} ({exception})")
-        logger.debug(self._streamer.stats)
+        self._streamer = ThreadedIterableStreamer(stats_iterable, initial_stream_value={})
 
     def stop(self):
         self._streamer.stop()
 
-    @property
-    def stats(self):
+    def get_streamed_stats(self) -> Dict[str, Any]:
         stats = self._streamer.stats
         if stats["Error"]:
-            self._log_debug("Stats fetching failed", stats["Error"])
+            logger.error(f"containers (Podman) Container({self._container.id}): Stats fetching failed")
+            logger.debug(f"containers (Podman) Container({self._container.id}): ", stats)
 
         return stats["Stats"][0]
 
     @property
-    def activity_stats(self):
-        result_stats = {"cpu": {}, "memory": {}, "io": {}, "network": {}}
-        api_stats = self.stats
+    def activity_stats(self) -> Dict[str, Any]:
+        """Activity Stats
 
-        if any(field not in api_stats for field in self.MANDATORY_FIELDS):
-            self._log_debug("Missing mandatory fields")
-            return result_stats
+        Each successive access of activity_stats will cause computation of activity_stats
+        """
+        computed_activity_stats = self._compute_activity_stats()
+        self._old_computed_stats = computed_activity_stats
+        self._last_stats_computed_time = time.time()
+        return computed_activity_stats
+
+    def _compute_activity_stats(self) -> Dict[str, Dict[str, Any]]:
+        stats = {"cpu": {}, "memory": {}, "io": {}, "network": {}}
+        api_stats = self.get_streamed_stats()
+
+        if any(field not in api_stats for field in self.MANDATORY_FIELDS) or (
+            "Network" not in api_stats and any(k not in api_stats for k in ['NetInput', 'NetOutput'])
+        ):
+            logger.error(f"containers (Podman) Container({self._container.id}): Missing mandatory fields")
+            return stats
 
         try:
-            cpu_usage = float(api_stats.get("CPU", 0))
+            stats["cpu"]["total"] = api_stats['CPU']
 
-            mem_usage = float(api_stats["MemUsage"])
-            mem_limit = float(api_stats["MemLimit"])
+            stats["memory"]["usage"] = api_stats["MemUsage"]
+            stats["memory"]["limit"] = api_stats["MemLimit"]
 
-            rx = float(api_stats["NetInput"])
-            tx = float(api_stats["NetOutput"])
+            stats["io"]["ior"] = api_stats["BlockInput"]
+            stats["io"]["iow"] = api_stats["BlockOutput"]
+            stats["io"]["time_since_update"] = 1
+            # Hardcode `time_since_update` to 1 as podman already sends at the same fixed rate per second
 
-            ior = float(api_stats["BlockInput"])
-            iow = float(api_stats["BlockOutput"])
+            if "Network" not in api_stats:
+                # For podman rooted mode
+                stats["network"]['rx'] = api_stats["NetInput"]
+                stats["network"]['tx'] = api_stats["NetOutput"]
+                stats["network"]['time_since_update'] = 1
+                # Hardcode to 1 as podman already sends at the same fixed rate per second
+            else:
+                # For podman in rootless mode
+                # Note: No stats are being sent as of now in rootless mode. They are defaulting to 0 on podman end.
+                stats['network'] = {
+                    "cumulative_rx": sum(interface["RxBytes"] for interface in api_stats["Network"].values()),
+                    "cumulative_tx": sum(interface["TxBytes"] for interface in api_stats["Network"].values()),
+                }
+                # Using previous stats to calculate rates
+                old_network_stats = self._old_computed_stats.get("network")
+                if old_network_stats:
+                    stats['network']['time_since_update'] = round(self.time_since_update)
+                    stats['network']['rx'] = stats['network']['cumulative_rx'] - old_network_stats["cumulative_rx"]
+                    stats['network']['tx'] = stats['network']['cumulative_tx'] - old_network_stats['cumulative_tx']
 
-            # Hardcode `time_since_update` to 1 as podman
-            # already sends the calculated rate per second
-            result_stats = {
-                "cpu": {"total": cpu_usage},
-                "memory": {"usage": mem_usage, "limit": mem_limit},
-                "io": {"ior": ior, "iow": iow, "time_since_update": 1},
-                "network": {"rx": rx, "tx": tx, "time_since_update": 1},
-            }
         except ValueError as e:
-            self._log_debug("Non float stats values found", e)
+            logger.error(f"containers (Podman) Container({self._container.id}): Non float stats values found", e)
 
-        return result_stats
+        return stats
+
+    @property
+    def time_since_update(self) -> float:
+        # In case no update (at startup), default to 1
+        return max(1, self._streamer.last_update_time - self._last_stats_computed_time)
 
 
 class PodmanPodStatsFetcher:
@@ -92,7 +125,7 @@ class PodmanPodStatsFetcher:
         # Threaded Streamer
         # Temporary patch to get podman extension working
         stats_iterable = (pod_manager.stats(decode=True) for _ in iter(int, 1))
-        self._streamer = StatsStreamer(stats_iterable, initial_stream_value={}, sleep_duration=2)
+        self._streamer = ThreadedIterableStreamer(stats_iterable, initial_stream_value={}, sleep_duration=2)
 
     def _log_debug(self, msg, exception=None):
         logger.debug(f"containers (Podman): Pod Manager - {msg} ({exception})")
@@ -118,13 +151,13 @@ class PodmanPodStatsFetcher:
                 "io": io_stats or {},
                 "memory": memory_stats or {},
                 "network": network_stats or {},
-                "cpu": cpu_stats or {"total": 0.0},
+                "cpu": cpu_stats or {},
             }
             result_stats[stat["CID"]] = computed_stats
 
         return result_stats
 
-    def _get_cpu_stats(self, stats):
+    def _get_cpu_stats(self, stats: Dict) -> Optional[Dict]:
         """Return the container CPU usage.
 
         Output: a dict {'total': 1.49}
@@ -136,7 +169,7 @@ class PodmanPodStatsFetcher:
         cpu_usage = string_value_to_float(stats["CPU"].rstrip("%"))
         return {"total": cpu_usage}
 
-    def _get_memory_stats(self, stats):
+    def _get_memory_stats(self, stats) -> Optional[Dict]:
         """Return the container MEMORY.
 
         Output: a dict {'usage': ..., 'limit': ...}
@@ -157,7 +190,7 @@ class PodmanPodStatsFetcher:
 
         return {'usage': usage, 'limit': limit, 'inactive_file': 0}
 
-    def _get_network_stats(self, stats):
+    def _get_network_stats(self, stats) -> Optional[Dict]:
         """Return the container network usage using the Docker API (v1.0 or higher).
 
         Output: a dict {'time_since_update': 3000, 'rx': 10, 'tx': 65}.
@@ -180,10 +213,10 @@ class PodmanPodStatsFetcher:
             self._log_debug("Compute MEM usage failed", e)
             return None
 
-        # Hardcode `time_since_update` to 1 as podman docs don't specify the rate calculated procedure
+        # Hardcode `time_since_update` to 1 as podman docs don't specify the rate calculation procedure
         return {"rx": rx, "tx": tx, "time_since_update": 1}
 
-    def _get_io_stats(self, stats):
+    def _get_io_stats(self, stats) -> Optional[Dict]:
         """Return the container IO usage using the Docker API (v1.0 or higher).
 
         Output: a dict {'time_since_update': 3000, 'ior': 10, 'iow': 65}.
@@ -206,7 +239,7 @@ class PodmanPodStatsFetcher:
             self._log_debug("Compute BlockIO usage failed", e)
             return None
 
-        # Hardcode `time_since_update` to 1 as podman docs don't specify the rate calculated procedure
+        # Hardcode `time_since_update` to 1 as podman docs don't specify the rate calculation procedure
         return {"ior": ior, "iow": iow, "time_since_update": 1}
 
 
@@ -242,7 +275,7 @@ class PodmanContainersExtension:
         # return self.client.version()
         return {}
 
-    def stop(self):
+    def stop(self) -> None:
         # Stop all streaming threads
         for t in itervalues(self.container_stats_fetchers):
             t.stop()
@@ -250,7 +283,7 @@ class PodmanContainersExtension:
         if self.pods_stats_fetcher:
             self.pods_stats_fetcher.stop()
 
-    def update(self, all_tag):
+    def update(self, all_tag) -> Tuple[Dict, list[Dict[str, Any]]]:
         """Update Podman stats using the input method."""
 
         if not self.client:
@@ -298,55 +331,58 @@ class PodmanContainersExtension:
         return version_stats, container_stats
 
     @property
-    def key(self):
+    def key(self) -> str:
         """Return the key of the list."""
         return 'name'
 
-    def generate_stats(self, container):
+    def generate_stats(self, container) -> Dict[str, Any]:
         # Init the stats for the current container
         stats = {
             'key': self.key,
-            # Export name
             'name': nativestr(container.name),
-            # Container Id
             'id': container.id,
-            # Container Image
             'image': ','.join(container.image.tags if container.image.tags else []),
-            # Container Status (from attrs)
             'status': container.attrs['State'],
             'created': container.attrs['Created'],
             'command': container.attrs.get('Command') or [],
+            'io': {},
+            'cpu': {},
+            'memory': {},
+            'network': {},
+            'io_rx': None,
+            'io_wx': None,
+            'cpu_percent': None,
+            'memory_percent': None,
+            'network_rx': None,
+            'network_tx': None,
+            'uptime': None,
         }
 
-        if stats['status'] in self.CONTAINER_ACTIVE_STATUS:
-            started_at = datetime.fromtimestamp(container.attrs['StartedAt'])
-            stats_fetcher = self.container_stats_fetchers[container.id]
-            activity_stats = stats_fetcher.activity_stats
-            stats.update(activity_stats)
+        if stats['status'] not in self.CONTAINER_ACTIVE_STATUS:
+            return stats
 
-            # Additional fields
-            stats['cpu_percent'] = stats["cpu"]['total']
-            stats['memory_usage'] = stats["memory"].get('usage')
-            if stats['memory'].get('cache') is not None:
-                stats['memory_usage'] -= stats['memory']['cache']
-            stats['io_rx'] = stats['io'].get('ior') // stats['io'].get('time_since_update')
-            stats['io_wx'] = stats['io'].get('iow') // stats['io'].get('time_since_update')
-            stats['network_rx'] = stats['network'].get('rx') // stats['network'].get('time_since_update')
-            stats['network_tx'] = stats['network'].get('tx') // stats['network'].get('time_since_update')
-            stats['uptime'] = pretty_date(started_at)
-            # Manage special chars in command (see isse#2733)
-            stats['command'] = replace_special_chars(' '.join(stats['command']))
-        else:
-            stats['io'] = {}
-            stats['cpu'] = {}
-            stats['memory'] = {}
-            stats['network'] = {}
-            stats['io_rx'] = None
-            stats['io_wx'] = None
-            stats['cpu_percent'] = None
-            stats['memory_percent'] = None
-            stats['network_rx'] = None
-            stats['network_tx'] = None
-            stats['uptime'] = None
+        stats_fetcher = self.container_stats_fetchers[container.id]
+        activity_stats = stats_fetcher.activity_stats
+        stats.update(activity_stats)
+
+        # Additional fields
+        stats['cpu_percent'] = stats['cpu'].get('total')
+        stats['memory_usage'] = stats['memory'].get('usage')
+        if stats['memory'].get('cache') is not None:
+            stats['memory_usage'] -= stats['memory']['cache']
+
+        if all(k in stats['io'] for k in ('ior', 'iow', 'time_since_update')):
+            stats['io_rx'] = stats['io']['ior'] // stats['io']['time_since_update']
+            stats['io_wx'] = stats['io']['iow'] // stats['io']['time_since_update']
+
+        if all(k in stats['network'] for k in ('rx', 'tx', 'time_since_update')):
+            stats['network_rx'] = stats['network']['rx'] // stats['network']['time_since_update']
+            stats['network_tx'] = stats['network']['tx'] // stats['network']['time_since_update']
+
+        started_at = datetime.fromtimestamp(container.attrs['StartedAt'])
+        stats['uptime'] = pretty_date(started_at)
+
+        # Manage special chars in command (see issue#2733)
+        stats['command'] = replace_special_chars(' '.join(stats['command']))
 
         return stats

--- a/glances/plugins/containers/stats_streamer.py
+++ b/glances/plugins/containers/stats_streamer.py
@@ -11,11 +11,11 @@ import time
 from glances.logger import logger
 
 
-class StatsStreamer:
+class ThreadedIterableStreamer:
     """
     Utility class to stream an iterable using a background / daemon Thread
 
-    Use `StatsStreamer.stats` to access the latest streamed results
+    Use `ThreadedIterableStreamer.stats` to access the latest streamed results
     """
 
     def __init__(self, iterable, initial_stream_value=None, sleep_duration=0.1):

